### PR TITLE
Issue 7468 - RFE - HIBP response caching

### PR DESCRIPTION
**Pull request by jchapma**
Description:
Add LRU cache for HIBP API responses to reduce latency and external API calls. The cache stores range query responses (prefix -> suffix list) rather than individual password verdicts, providing implicit negative caching since a cache hit returns the complete suffix list for matching.

- Hash table (256 buckets) plus a doubly linked list for LRU eviction
- Configurable size (default: 1000 entries) and TTL (default: 24 hours)
- Concurrency protection using NSPR locks
- Cache statistics tracking (hits/misses)
- Unit tests for cache operations

Dependencies:
- Issue 7468 - RFE - Add HIBP HTTP client
- Issue 7468 - RFE - HIBP password breach validation

Relates: https://github.com/389ds/389-ds-base/issues/7468

Reviewed by:

## Summary by Sourcery

Introduce an in-memory LRU cache for Have I Been Pwned (HIBP) range query responses and integrate it into the HIBP client lifecycle.

New Features:
- Add configurable, TTL-based LRU cache for HIBP prefix responses with hit/miss statistics.
- Cache HIBP API responses in hibp_query_api to reduce repeated external calls and latency.

Enhancements:
- Wire HIBP subsystem shutdown into the main server cleanup path to release cache resources cleanly.
- Refine HIBP public API header to separate core API, hash provider abstraction, and cache/test helper interfaces.

Tests:
- Add unit tests covering HIBP cache basic operations, LRU eviction behavior, TTL expiration, and LRU reordering under access patterns.

### CODE REVIEW

File: ldap/servers/slapd/hibp_client.c [hunk 1]
```diff
@@ -392,6 +746,13 @@ hibp_query_api(const char *prefix, const char *api_url, HIBPResponse *response,
         return -1;
     }
 
+    /* Store successful response in cache */
+    if (g_hibp_cache && response->data && response->size > 0) {
+        hibp_cache_put(prefix, response->data, response->size);
```
Comment (droideck) [inline ldap/servers/slapd/hibp_client.c:751]: Will you additionally modify the PR once another one is merged?
Currently, cache key ignores the configured breach DB URL which is understandable... But probably it should be handled as passwordBreachDbUrl can differ by policy, including local policies. A response from the public HIBP endpoint can be reused for an internal/custom endpoint with the same prefix, or vice versa, producing false "safe" or false "breached" decisions.

File: ldap/servers/slapd/hibp_client.c [hunk 2]
```diff
@@ -80,6 +82,341 @@ typedef struct hibp_response
     size_t capacity;
 } HIBPResponse;
 
+/*
+ * LRU Cache for HIBP API responses
+ *
+ * Caches API responses by hash prefix.
+ * Uses double linked list for LRU ordering.
+ */
+#define HIBP_CACHE_DEFAULT_SIZE 1000    /* Default max cache entries */
```
Comment (droideck) [inline ldap/servers/slapd/hibp_client.c:91]: IIUC, each accepted response can be up to HIBP_RESPONSE_MAX_SIZE of 1 MB. That makes the theoretical default cache size about 1 GB. That's a lot... At least, I think, we need to document it very explicitly. And maybe let it be configured? And even a way to disable the cache?

File: ldap/servers/slapd/hibp.h [hunk 3]
```diff
@@ -11,39 +11,80 @@
...
+
+/* Destroy cache and free all resources */
+void hibp_cache_destroy(void);
+
+/*
+ * Get cached response for prefix
+ *
+ * Returns: response data (caller must free), NULL if not cached
+ */
+char *hibp_cache_get(const char *prefix, size_t *response_size);
```
Comment (droideck) [inline ldap/servers/slapd/hibp.h:66]: Do we really need it to be public? 
Probably all of them can be left as static in hibp_client.c:
``` 
hibp_cache_get
hibp_cache_put
hibp_cache_init
hibp_cache_destroy
```

File: ldap/servers/slapd/hibp_client.c [hunk 4]
```diff
@@ -322,6 +659,23 @@ hibp_query_api(const char *prefix, const char *api_url, HIBPResponse *response,
         return 0;
     }
 
+    /* Check cache first */
+    if (g_hibp_cache) {
+        size_t cached_size;
+        char *cached_data = hibp_cache_get(prefix, &cached_size);
+        if (cached_data) {
+            slapi_log_err(SLAPI_LOG_INFO, "hibp_query_api",
+                          "Cache HIT for prefix %s (size=%zu bytes)\n", prefix, cached_size);
```
Comment (droideck) [inline ldap/servers/slapd/hibp_client.c:668]: A few slapi_log_err calls log every prefix hit/miss/store at INFO. Yeah, the prefix is intentionally partial, but it is still password-derived material... I’d downgrade this to debug/trace and avoid logging the prefix by default.

**Review by droideck — CHANGES_REQUESTED** [review]
The new unit tests exercise bare cache operations, but not the real hibp_check_password() cache path because hibp_query_api() returns mock responses before cache lookup.
I think that would be nice to add tests for cache hit/miss through hibp_check_password, URL-scoped cache keys, duplicate prefix across two URLs, oversized response cache skipping/eviction, and concurrent miss storms?.. And run some ASAN with it, probably.

File: ldap/servers/slapd/hibp_client.c (hunk 1 above)
Comment (jchapma) [reply ldap/servers/slapd/hibp_client.c:751]: Good point, I made some updates so each cache entry stores the api url it came from, and lookups match on both (prefix, url). Now responses from different endpoints wont conflict with each other

File: ldap/servers/slapd/hibp_client.c (hunk 2 above)
Comment (jchapma) [reply ldap/servers/slapd/hibp_client.c:91]: Yes, in theory the default cache size could grow to 1GB, but in practice, real responses are ~40KB, giving a default size of ~40MB. But, I made some updates that will allow the admin tune or disable caching

File: ldap/servers/slapd/hibp.h (hunk 3 above)
Comment (jchapma) [reply ldap/servers/slapd/hibp.h:66]: We dont, I made all four static

File: ldap/servers/slapd/hibp_client.c (hunk 4 above)
Comment (jchapma) [reply ldap/servers/slapd/hibp_client.c:668]: Downgraded cache hit/miss/store logging from INFO to DEBUG and removed the prefix from all log messages

**Comment by droideck** [comment]
In the comments, you said you made the changes but I can't see them... Were they pushed? Or we might have some weird GH bug)
